### PR TITLE
Running setup.py directly is deprecated

### PR DESCRIPTION
**Pull request by cclauss**
Running `python3 setup.py` directly as a script is deprecated.
* https://setuptools.pypa.io/en/latest/deprecated/commands.html
* https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html#summary

Let's use [`python3 -m build`](https://build.pypa.io) and `python3 -m pip install` instead of directly running `setup.py`.

### CODE REVIEW

**Review by Nusnus — APPROVED** [review]
LGTM

**Comment by cclauss** [comment]
This PR is `draft` because I need to do `Makefile` but will do that in the next few hours.

File: README.rst
Comment (ulgens) [inline README.rst:268]: I'm not sure venv management should be covered here. There are tons of ways of to manage a venv, and forcing one in a tutorial ends up badly in my experience. Just a mention of a venv should be enough, as is before.

**Review by ulgens — CHANGES_REQUESTED** [review]

File: docs/includes/installation.txt
Comment (ulgens) [inline docs/includes/installation.txt:14]: Same here: https://github.com/celery/django-celery-beat/pull/737#discussion_r1503724648 With these changes, it looks like the steps are more about venvs, not the actual goal of the doc.

File: t/unit/conftest.py
Comment (ulgens) [inline t/unit/conftest.py:5]: Doesn't it make better sense with "yet" at end of the sentence, as the before?

**Review by ulgens — CHANGES_REQUESTED** [review]

File: README.rst
Comment (cclauss) [reply README.rst:268]: This is Python's built-in way of creating and enabling a venv.  Users are free to choose alternative methods but I prefer explicit steps here to document what we know works and avoid issues being opened around implicit steps.

File: t/unit/conftest.py
Comment (cclauss) [reply t/unit/conftest.py:5]: It felt like a dangling participle when I read it.  (But I would bet dollars to doughnuts that it is not a participle. 😃 )

File: Makefile
Comment (deronnax) [inline Makefile:69]: that direct invocation of setup.py should also be removed, no?

Comment (cclauss) [reply Makefile:69]: Discussed at https://github.com/celery/django-celery-beat/pull/737#issuecomment-1966276680

Comment (deronnax) [reply Makefile:69]: Ah, I did not make the link, sorry.

**Comment by deronnax** [comment]
I went to look into setuptools' changelog what these `register` and `upload --sign` commands that I never saw before were, and setuptools tells us you can forget about them and just do with twine https://github.com/pypa/setuptools/pull/1898.
So I think you can just remove these two lines. If you want, you can test locally that it still work as you expect by making twine point to the test version of PyPI: https://test.pypi.org/

**Review by Nusnus — APPROVED** [review]
LGTM

File: Makefile
Comment (deronnax) [inline Makefile:72]:
```suggestion
	twine upload --sign --identity="$(PGPIDENT) dist/*"
```

File: Makefile
Comment (deronnax) [inline Makefile:69]:
```suggestion
```
You don't need to register at all anymore on PyPI now.

File: README.rst
Comment (cclauss) [reply README.rst:268]: Opinionated tooling would have been a suggestion to use `uv`. 😄

**Review by deronnax — CHANGES_REQUESTED** [review]
I checked out the PR to test and I got a problem:
```
make
Makefile:68: *** missing separator.  Stop.
```
On line 68, you used space for indentation instead of tabs. Makefile *only* accepts tabs for indentation.

File: Makefile
Comment (amweiss) [inline Makefile:68]:
```suggestion
	python -m pip install --upgrade build setuptools wheel twine
```

**Review by amweiss — CHANGES_REQUESTED** [review]

**Comment by amweiss** [comment]
Even with the above fixes, the steps in the README do not work for me locally:

```
 => [base] exporting to image                                                                                                                                                  0.0s
 => => exporting layers                                                                                                                                                        0.0s
 => => writing image sha256:5fe8d8101eb68554633bd1a71f3197d691c35206b3fe0a4bf7254a8fe3a3d0ab                                                                                   0.0s
 => => naming to docker.io/library/django-celery-beat-base                                                                                                                     0.0s
 => [celery-beat internal] load .dockerignore                                                                                                                                  0.0s
 => => transferring context: 2B                                                                                                                                                0.0s
 => [celery-beat internal] load build definition from Dockerfile                                                                                                               0.1s
 => => transferring dockerfile: 325B                                                                                                                                           0.0s
 => ERROR [celery-beat internal] load metadata for docker.io/library/django-celery-beat_base:latest                                                                            2.4s
 => [django internal] load .dockerignore                                                                                                                                       0.0s
 => => transferring context: 2B                                                                                                                                                0.0s
 => [django internal] load build definition from Dockerfile                                                                                                                    0.1s
 => => transferring dockerfile: 315B                                                                                                                                           0.0s
 => [celery-beat auth] library/django-celery-beat_base:pull token for registry-1.docker.io                                                                                     0.0s
------
 > [celery-beat internal] load metadata for docker.io/library/django-celery-beat_base:latest:
------
failed to solve: django-celery-beat_base: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

However, this is also true on the `main` branch as well as `v2.5.0` at this point, so I am unsure what the best path forward would be.

**Comment by deronnax** [comment]
What you are posting is seemingly a docker error, and there is no docker in the Makefile, so I am not sure how you got it. For the Makefile, I was able to have `make release` to work, that's interests us here.

Regarding the docker error, it's something trivially wrong in the dockerfiles: the Dockerfiles ask for a tagged base image that is never generated, hence the error. It probably never worked.

**Comment by cclauss** [comment]
@deronnax 
> I was able to have `make release` work, that's interests us here.
> [ ... ] 
> [Docker] probably never worked.

@Nusnus I agree with these statements and am ready to proceed with this PR as-is.

**Comment by amweiss** [comment]
> What you are posting is seemingly a docker error, and there is no docker in the Makefile, so I am not sure how you got it. For the Makefile, I was able to have `make release` to work, that's what interests us here.
> 
> Regarding the docker error, it's something trivially wrong in the dockerfiles: the Dockerfiles ask for a tagged base image that is never generated, hence the error. It probably never worked.

This PR changed the Dockerfile, so I wanted to see what it did. However, it seems like it's never worked.

File: docker/base/Dockerfile
Comment (deronnax) [inline docker/base/Dockerfile:24]:
```suggestion
RUN pip3 install --user --editable ".[develop]"
```

**Review by deronnax — APPROVED** [review]

**Comment by auvipy** [comment]
as the oldest maintainer of the project, I would like to review it in detail before merging. hope you are OK with it.

**Comment by Nusnus** [comment]
@auvipy 
> as the oldest maintainer of the project, I would like to review it in detail before merging. hope you are OK with it.

Please be aware that the release for Django 5 support is blocked on this PR: #680

From my understanding this PR is also ready already, even if not everything is perfect.

Many are expecting a release soon, IMHO I believe we should proceed - let me know what I can do to accelerate the merge, how may I help?

P.S
I'm up for a zoom session or anything online to work together - let me know bro 🤙

**Comment by ulgens** [comment]
> Many are expecting a release soon, IMHO I believe we should proceed - let me know what I can do to accelerate the merge, how may I help?

I also agree that a new release with Django 5.0 support is needed, but I don't think this PR is a blocker.

**Comment by Nusnus** [comment]
> I also agree that a new release with Django 5.0 support is needed, but I don't think this PR is a blocker. 

Makes sense actually, reviewing again the changes of this PR.

If you agree @cclauss, we can move forward and continue with the PR after @auvipy gives us his feedback to proceed in both paths.

**Review by deronnax — APPROVED** [review]

**Comment by deronnax** [comment]
Looks good to me. Migrating the setup.py to a declarative form could be the next step, wdyt?

**Comment by amweiss** [comment]
With these changes in place, if you want to make the README work correctly or the docker compose actions work, the following changes are still needed. Add the following lines:
```
      tags:
        - django-celery-beat_base:latest
```
below https://github.com/cclauss/django-celery-beat/blob/bc4684b174d6d5c766dffc3a6f74241e9ca113f7/docker-compose.yml#L10-L11

so the `base` definition looks like the following:
```
  base:
    build:
      context: .
      dockerfile: docker/base/Dockerfile
      tags:
        - django-celery-beat_base:latest
    command: ["sleep", "inf"]
```

After that, the `docker-compose up --build` command will work as it's stated in the README.

**Review by amweiss — APPROVED** [review]
This worked for me now.

**Comment by cclauss** [comment]
This sat un-reviewed to too long until `setuptools` changes broke things in
* #771 
* #772 

Let's get some maintainers reviews on this please.

File: Makefile
Comment (dimbleby) [inline Makefile:68]: setuptools and wheel are not needed, `build` will create an isolated environment with build dependencies

**Review by thedrow — APPROVED** [review]

**Review by auvipy — COMMENTED** [review]
good job!

**Comment by browniebroke** [comment]
While this is fresh in everyone's head, would now be a good time to make a new release?  The change in setuptools was reverted, but it's been deprecated for years, so it will be removed at some point in... Would be nice to give plenty of time for upgrading. 

Thanks!

**Comment by cclauss** [comment]
😉 https://github.com/celery/django-celery-beat/pull/749#issuecomment-2256346238

**Pull request merged**
